In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# ---------------------------------
# ÉTAPE 1 : Importation des datasets
# ---------------------------------

paths = {
    "Matches_with_Elo": "../Donnees/Matches_with_Elo.csv",
    "Pred_Full": "../Donnees/matches_pred_full.csv",
    "Recent": "../Donnees/matches_recent.csv",
    "Clean": "../Donnees/matches_clean.csv",
    "Select_Championnats": "../Donnees/matches_select_championnats.csv"
}

datasets = {}

for name, path in paths.items():
    try:
        df = pd.read_csv(path, sep=";", parse_dates=["MatchDate"], low_memory=False)
    except:
        df = pd.read_csv(path, sep=",", parse_dates=["MatchDate"], low_memory=False)

    datasets[name] = df
    print(f"\n----- {name} -----")
    print(f"Shape : {df.shape}")
    print("Colonnes principales détectées :", df.columns[:10].tolist())
    print(df.head(2))


C:\Users\simon\AppData\Local\Temp\ipykernel_28264\3780388929.py:21: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df = pd.read_csv(path, sep=";", parse_dates=["MatchDate"], low_memory=False)



----- Matches_with_Elo -----
Shape : (165619, 66)
Colonnes principales détectées : ['Division', 'MatchDate', 'MatchTime', 'HomeTeam', 'AwayTeam', 'Form3Home', 'Form5Home', 'Form3Away', 'Form5Away', 'FTHome']
  Division  MatchDate MatchTime     HomeTeam      AwayTeam  Form3Home  \
0       T1 2003-08-17       NaN  Malatyaspor  A. Sebatspor        9.0   
1       T1 2003-08-31       NaN  Trabzonspor  A. Sebatspor        4.0   

   Form5Home  Form3Away  Form5Away  FTHome  ...  PDraw_norm PAway_norm  \
0       12.0        0.0        0.0     3.0  ...    0.236453   0.147783   
1        7.0        0.0        0.0     2.0  ...    0.205161   0.126907   

   ShotsDifference  CornersDifference CardsHome  CardsAway  CardsDiff  \
0              NaN                NaN       NaN        NaN        NaN   
1              NaN                NaN       NaN        NaN        NaN   

   GameDominanceIndex  Country  country.1  
0                 NaN      TUR        TUR  
1                 NaN      TUR        TU

C:\Users\simon\AppData\Local\Temp\ipykernel_28264\3780388929.py:21: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df = pd.read_csv(path, sep=";", parse_dates=["MatchDate"], low_memory=False)



----- Pred_Full -----
Shape : (165619, 48)
Colonnes principales détectées : ['Division', 'MatchDate', 'HomeTeam', 'AwayTeam', 'Division_1', 'MatchDate_2', 'HomeTeam_3', 'AwayTeam_4', 'Form3Home', 'Form5Home']
  Division  MatchDate     HomeTeam      AwayTeam Division_1 MatchDate_2  \
0       T1 2003-08-17  Malatyaspor  A. Sebatspor         T1  17/08/2003   
1       T1 2003-08-31  Trabzonspor  A. Sebatspor         T1  31/08/2003   

    HomeTeam_3    AwayTeam_4  Form3Home  Form5Home  ...  FormMomentumHome  \
0  Malatyaspor  A. Sebatspor        9.0       12.0  ...               6.0   
1  Trabzonspor  A. Sebatspor        4.0        7.0  ...               1.0   

   FormMomentumAway   ProbSum  PHome_norm  PDraw_norm  PAway_norm  Country  \
0               0.0  1.127778    0.615764    0.236453    0.147783      TUR   
1               0.0  1.125684    0.667932    0.205161    0.126907      TUR   

   country.1  Season  FTResult  
0        TUR    2003         H  
1        TUR    2003         H 

C:\Users\simon\AppData\Local\Temp\ipykernel_28264\3780388929.py:21: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df = pd.read_csv(path, sep=";", parse_dates=["MatchDate"], low_memory=False)


In [6]:

summary_rows = []

for name, df in datasets.items():
    
    # Taux NaN moyen
    nan_rate = df.isna().mean().mean()

    # Distribution FTResult
    if "FTResult" in df.columns:
        dist = df["FTResult"].value_counts(normalize=True).reindex(["H","D","A"]).fillna(0)
        h_rate, d_rate, a_rate = dist["H"], dist["D"], dist["A"]
    else:
        h_rate = d_rate = a_rate = np.nan

    # Range temporel
    if "MatchDate" in df.columns:
        date_min = df["MatchDate"].min()
        date_max = df["MatchDate"].max()
    else:
        date_min = date_max = None

    summary_rows.append({
        "Dataset": name,
        "Lignes": df.shape[0],
        "Colonnes": df.shape[1],
        "NaN moyen (%)": round(nan_rate * 100, 2),
        "Home (%)": round(h_rate * 100, 2),
        "Draw (%)": round(d_rate * 100, 2),
        "Away (%)": round(a_rate * 100, 2),
        "Date min": date_min,
        "Date max": date_max,
        "Complétude (%)": round((1 - nan_rate) * 100, 2)
    })

summary_table = pd.DataFrame(summary_rows)
summary_table


,Dataset,Lignes,Colonnes,NaN moyen (%),Home (%),Draw (%),Away (%),Date min,Date max,Complétude (%)
0,Matches_with_Elo,165619,66,18.26,45.16,26.69,28.15,2000-07-28 00:00:00,2025-02-26 00:00:00,81.74
1,Pred_Full,165619,48,7.76,45.16,26.69,28.15,2000-07-28 00:00:00,2025-02-26 00:00:00,92.24
2,Recent,138312,46,5.29,44.82,26.56,28.61,01/01/2006,31/12/2023,94.71
3,Clean,8478,35,1.53,43.49,26.88,29.63,01/01/2024,31/10/2024,98.47
4,Select_Championnats,52580,48,2.65,44.82,25.91,29.28,2006-01-01 00:00:00,2025-02-26 00:00:00,97.35


In [7]:
# ---------------------------------------
# ÉTAPE 3 : Harmonisation et nettoyage
# ---------------------------------------

# Features prédictives communes
FEATURES_BASE = [
    "HomeEloSnap", "AwayEloSnap", "EloDiff", "EloTotal", "EloAdvantage",
    "Form3Home", "Form5Home", "Form3Away", "Form5Away",
    "Form3Diff", "Form5Diff",
    "OddHome", "OddDraw", "OddAway",
    "PHome_norm", "PDraw_norm", "PAway_norm", "ProbSum"
]

cleaned_datasets = {}

for name, original_df in datasets.items():
    df = original_df.copy()
    
    print(f"\n{name}")

    # 1) Supprimer colonnes dupliquées (_1, _2, _3)
    df = df.loc[:, ~df.columns.str.contains(r"_\d+$")]

    # 2) Garder seulement les features présentes dans ce dataset
    avail_features = [c for c in FEATURES_BASE if c in df.columns]

    print(f"Features conservées ({len(avail_features)}):", avail_features)

    # 3) Filtrer les matchs avec résultat connu
    df = df[df["FTResult"].isin(["H", "D", "A"])]

    # 4) Nettoyer inf/-inf → NaN
    df[avail_features] = df[avail_features].replace([np.inf, -np.inf], np.nan)

    # 5) Imputation médiane (robuste aux outliers)
    for col in avail_features:
        df[col] = df[col].fillna(df[col].median())

    # 6) Conserver seulement les colonnes utiles
    cleaned = df[["MatchDate", "FTResult"] + avail_features].sort_values("MatchDate")
    cleaned = cleaned.reset_index(drop=True)

    cleaned_datasets[name] = cleaned

    print(f"→ Dataset final : {cleaned.shape}")



Matches_with_Elo
Features conservées (18): ['HomeEloSnap', 'AwayEloSnap', 'EloDiff', 'EloTotal', 'EloAdvantage', 'Form3Home', 'Form5Home', 'Form3Away', 'Form5Away', 'Form3Diff', 'Form5Diff', 'OddHome', 'OddDraw', 'OddAway', 'PHome_norm', 'PDraw_norm', 'PAway_norm', 'ProbSum']
→ Dataset final : (165617, 20)

Pred_Full
Features conservées (18): ['HomeEloSnap', 'AwayEloSnap', 'EloDiff', 'EloTotal', 'EloAdvantage', 'Form3Home', 'Form5Home', 'Form3Away', 'Form5Away', 'Form3Diff', 'Form5Diff', 'OddHome', 'OddDraw', 'OddAway', 'PHome_norm', 'PDraw_norm', 'PAway_norm', 'ProbSum']
→ Dataset final : (165617, 20)

Recent
Features conservées (18): ['HomeEloSnap', 'AwayEloSnap', 'EloDiff', 'EloTotal', 'EloAdvantage', 'Form3Home', 'Form5Home', 'Form3Away', 'Form5Away', 'Form3Diff', 'Form5Diff', 'OddHome', 'OddDraw', 'OddAway', 'PHome_norm', 'PDraw_norm', 'PAway_norm', 'ProbSum']
→ Dataset final : (138310, 20)

Clean
Features conservées (18): ['HomeEloSnap', 'AwayEloSnap', 'EloDiff', 'EloTotal', 'El

In [8]:
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import log_loss, accuracy_score, f1_score
import numpy as np
import pandas as pd
import warnings

warnings.filterwarnings("ignore")

LABELS = ["H", "D", "A"]

def make_time_splits(dates, n_splits=3, min_train_size=5000):
    """
    Splits temporels sécurisés pour datasets réindexés.
    """
    n = len(dates)
    fold_sizes = np.full(n_splits, n // n_splits, dtype=int)
    fold_sizes[: n % n_splits] += 1
    
    bounds = np.cumsum(fold_sizes)
    splits = []

    for i in range(n_splits - 1):
        val_start = bounds[i]
        val_end   = bounds[i+1]

        train_idx = np.arange(0, val_start)
        val_idx   = np.arange(val_start, val_end)

        if len(train_idx) >= min_train_size:
            splits.append((train_idx, val_idx))

    return splits



def multiclass_brier_score(y_true, proba, labels):
    lab2idx = {lab:i for i, lab in enumerate(labels)}
    y_idx = np.array([lab2idx[y] for y in y_true])
    
    y_onehot = np.zeros_like(proba)
    y_onehot[np.arange(len(y_true)), y_idx] = 1.0
    
    return np.mean((proba - y_onehot)**2)


# -----------------------------------------
# ENTRAÎNEMENT LOGISTIQUE SUR LES 5 DATASETS
# -----------------------------------------

results_all = []

for name, df in cleaned_datasets.items():

    print(f"\n=========== {name} ===========")
    
    features = [c for c in df.columns if c not in ["MatchDate", "FTResult"]]
    
    splits = make_time_splits(df["MatchDate"], n_splits=3, min_train_size=5000)
    print(f"{len(splits)} splits temporels générés")

    # pipeline = imputation déjà faite → scaler + logreg
    preproc = ColumnTransformer([
        ("num", Pipeline([("scaler", StandardScaler())]), features)
    ])

    model = LogisticRegression(
        multi_class="multinomial",
        solver="lbfgs",
        C=1.0,
        max_iter=300,
        random_state=42
    )

    records = []

    for fold_id, (tr_idx, va_idx) in enumerate(splits, start=1):
        X_train = df.loc[tr_idx, features]
        y_train = df.loc[tr_idx, "FTResult"]

        X_val = df.loc[va_idx, features]
        y_val = df.loc[va_idx, "FTResult"]

        pipe = Pipeline([
            ("preproc", preproc),
            ("clf", model)
        ])

        pipe.fit(X_train, y_train)
        proba_val = pipe.predict_proba(X_val)
        y_pred = pipe.predict(X_val)

        # Metrics
        ll = log_loss(y_val, proba_val, labels=LABELS)
        acc = accuracy_score(y_val, y_pred)
        f1m = f1_score(y_val, y_pred, average="macro")
        brier = multiclass_brier_score(y_val, proba_val, LABELS)

        records.append({
            "Dataset": name,
            "Fold": fold_id,
            "LogLoss": ll,
            "Accuracy": acc,
            "F1_macro": f1m,
            "Brier": brier
        })

        print(f"[Fold {fold_id}] LogLoss={ll:.4f} | Acc={acc:.3f} | F1={f1m:.3f} | Brier={brier:.4f}")

    results_all.append(pd.DataFrame(records))


# Résumé global
results_all = pd.concat(results_all, ignore_index=True)
summary = results_all.groupby("Dataset").agg(["mean", "std"]).round(4)

print("\n RESUME ")  
print(summary)



=========== Matches_with_Elo ===========
2 splits temporels générés
[Fold 1] LogLoss=1.0009 | Acc=0.503 | F1=0.363 | Brier=0.2804
[Fold 2] LogLoss=0.9942 | Acc=0.511 | F1=0.385 | Brier=0.2793

=========== Pred_Full ===========
2 splits temporels générés
[Fold 1] LogLoss=1.0009 | Acc=0.503 | F1=0.363 | Brier=0.2804
[Fold 2] LogLoss=0.9942 | Acc=0.511 | F1=0.385 | Brier=0.2793

=========== Recent ===========
2 splits temporels générés
[Fold 1] LogLoss=0.9981 | Acc=0.507 | F1=0.372 | Brier=0.2779
[Fold 2] LogLoss=0.9972 | Acc=0.507 | F1=0.371 | Brier=0.2778

=========== Clean ===========
1 splits temporels générés
[Fold 1] LogLoss=1.0000 | Acc=0.501 | F1=0.377 | Brier=0.2804

=========== Select_Championnats ===========
2 splits temporels générés
[Fold 1] LogLoss=0.9878 | Acc=0.517 | F1=0.378 | Brier=0.2848
[Fold 2] LogLoss=0.9853 | Acc=0.521 | F1=0.387 | Brier=0.2818

 RESUME 
                    Fold         LogLoss         Accuracy         F1_macro  \
                    mean     std  

1) Select_Championnats :
    * Accuracy la plus élevée : 0.52
    * Logloss la plus faible : 0.985
